### ISSUE TITLE
Bug in 'par_fluxcalib' that affects the sensfunc function

### ISSUE THREAD

**Issue by Codeprey** [pypeit/PypeIt#1791]
Hello, 

I am using the pypeit v1.15.0 to reduce P200 DBSP spectra, and I encounter a bug in fitting the standard star sensitivity curve. 
The bug is shown as follows:

Input:
pypeit_sensfunc Science/spec1d_blue0059-G191B2B_DBSPb.fits -o std_g191b2b_sens.fits --algorithm UVIS

Output:
WARNING: VerifyWarning: Keyword name 'DISPANGLE' is greater than 8 characters or contains characters not allowed by the FITS standard; a HIERARCH card will be created. [astropy.io.fits.card]
[INFO]    :: Writing the parameters to sensfunc.par
[WARNING] :: Selected configuration file already exists and will be overwritten!
[INFO]    :: Getting archival standard spectrum
[INFO]    :: Loading standard star file: /opt/anaconda3/envs/pypeit/lib/python3.11/site-packages/pypeit/data/standards/calspec/g191b2b_stisnic_002.fits.gz
[INFO]    :: Fluxes are flambda, normalized to 1e-17
[INFO]    :: Applying extinction correction
[INFO]    :: Using palomarextinct.dat for extinction corrections.
[WARNING] :: Extrapolating at low wavelengths using last valid value
[INFO]    ::  Masking bad pixels
[INFO]    ::  Masking edges
[INFO]    ::  Masking Below the atmospheric cutoff
[INFO]    :: Masking hydrogen recombination lines
[INFO]    ::  Masking Balmer
[INFO]    ::  Masking Paschen
[INFO]    ::  Masking Brackett
[INFO]    ::  Masking Pfund
[INFO]    :: Your spectrum is bluer than 9100A, only optical telluric regions are masked.
[WARNING] :: Maximum number of iterations maxiter=10 reached in robust_polyfit_djs
[INFO]    :: Replacing bspline fit with polyfit over Hydrogen Recombination line regions
[INFO]    ::  This is the standard pixel: 1.08 Å
[INFO]    ::  This is the standard resolution element: 4.06 Å
[INFO]    ::  Breakpoint spacing: 20.32 pixels
[INFO]    :: Initialize bspline for flux calibration
[INFO]    :: Bspline fit on zeropoint. 
Traceback (most recent call last):
  File "/opt/anaconda3/envs/pypeit/bin/pypeit_sensfunc", line 8, in <module>
    sys.exit(SensFunc.entry_point())
             ^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda3/envs/pypeit/lib/python3.11/site-packages/pypeit/scripts/scriptbase.py", line 113, in entry_point
    cls.main(cls.parse_args())
  File "/opt/anaconda3/envs/pypeit/lib/python3.11/site-packages/pypeit/scripts/sensfunc.py", line 185, in main
    sensobj.run()
  File "/opt/anaconda3/envs/pypeit/lib/python3.11/site-packages/pypeit/sensfunc.py", line 452, in run
    self.flux_std()
  File "/opt/anaconda3/envs/pypeit/lib/python3.11/site-packages/pypeit/sensfunc.py", line 466, in flux_std
    self.sobjs_std.apply_flux_calib(self.par_fluxcalib, self.spectrograph, self)
  File "/opt/anaconda3/envs/pypeit/lib/python3.11/site-packages/pypeit/specobjs.py", line 576, in apply_flux_calib
    if par['extinct_correct'] is None else par['extinct_correct']
       ~~~^^^^^^^^^^^^^^^^^^^
TypeError: 'NoneType' object is not subscriptable


The problem seems to be that the 'par_fluxcalib' parameter is None. I've checked the corresponding scripts, if 'par_fluxcalib' is none it should be reset as the default value. (In some cases the default value seems to be none, not quite sure.) Before the recent update, I used v1.6.0 (in another conda environment) and this parameter was not introduced in 'sensfunc' functions, this command is smoothly run. Is there any bug in the definition of the 'par_fluxcalib' parameter? 

Thanks in advance,
Peter

**Comment by debora-pe** [comment]
Hi @Codeprey, what you understood is correct.  When `par_fluxcalib` is  `None` the code uses the default parameters. But `extinct_correct` is `None` by default. We need to decide how to best fix this, but in the meantime, you can create a sensitivity function configuration file, e.g.,  `sens_cfg.sens`, and add in the file the sensitivity function parameters, but also the fluxing parameters, e.g.,:

```
[sensfunc]
    algorithm = UVIS

[fluxcalib]
   extinct_correct = True
```

and then run `pypeit_sensfunc spec1dfile -s sens_cfg.sens`

I hope this helps.

**Comment by Codeprey** [comment]
Thanks Debora, following your suggestion I have done a test. However, the pypeit_sensfunc process automatically modifies the sens file and deletes the new added [fluxcalib] option. So the problem still exists.

**Comment by debora-pe** [comment]
@Codeprey can you confirm that you run `pypeit_sensfunc` with the `-s` flag? With this flag the sens file should not be modified. Actually, this is what I think happened.
When you add the `--algorithm UVIS` flag in your `pypeit_sensfunc` call, it overrides what it's written in the `sens_cfg.sens` file (apparently it overrides everything.)


I think in your case the call should be:
`pypeit_sensfunc Science/spec1d_blue0059-G191B2B_DBSPb.fits -o std_g191b2b_sens.fits -s sens_cfg.sens`

and the `sens_cfg.sens` file would have:
```
[sensfunc]
    algorithm = UVIS

[fluxcalib]
   extinct_correct = True
```

**Comment by Codeprey** [comment]
I've created a sensfunc.cfg file:

[sensfunc]
    algorithm = UVIS

[fluxcalib]
   extinct_correct = True


Then I delete the sensfunc.par file and run:
pypeit_sensfunc ./Science/spec1d_blue0059-G191B2B_DBSPb.fits -o std_sens.fits -s sensfunc.cfg


Output:
WARNING: VerifyWarning: Keyword name 'DISPANGLE' is greater than 8 characters or contains characters not allowed by the FITS standard; a HIERARCH card will be created. [astropy.io.fits.card]
[INFO]    :: Loading the reduction file
[INFO]    :: Sens input file loaded successfully.
[INFO]    :: Writing the parameters to sensfunc.par
[INFO]    :: Getting archival standard spectrum
[INFO]    :: Loading standard star file: /opt/anaconda3/envs/pypeit/lib/python3.11/site-packages/pypeit/data/standards/calspec/g191b2b_stisnic_002.fits.gz
[INFO]    :: Fluxes are flambda, normalized to 1e-17
[INFO]    :: Applying extinction correction
[INFO]    :: Using palomarextinct.dat for extinction corrections.
[WARNING] :: Extrapolating at low wavelengths using last valid value
[INFO]    ::  Masking bad pixels
[INFO]    ::  Masking edges
[INFO]    ::  Masking Below the atmospheric cutoff
[INFO]    :: Masking hydrogen recombination lines
[INFO]    ::  Masking Balmer
[INFO]    ::  Masking Paschen
[INFO]    ::  Masking Brackett
[INFO]    ::  Masking Pfund
[INFO]    :: Your spectrum is bluer than 9100A, only optical telluric regions are masked.
[WARNING] :: Maximum number of iterations maxiter=10 reached in robust_polyfit_djs
[INFO]    :: Replacing bspline fit with polyfit over Hydrogen Recombination line regions
[INFO]    ::  This is the standard pixel: 1.08 Å
[INFO]    ::  This is the standard resolution element: 4.06 Å
[INFO]    ::  Breakpoint spacing: 20.32 pixels
[INFO]    :: Initialize bspline for flux calibration
[INFO]    :: Bspline fit on zeropoint. 
Traceback (most recent call last):
  File "/opt/anaconda3/envs/pypeit/bin/pypeit_sensfunc", line 8, in <module>
    sys.exit(SensFunc.entry_point())
             ^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda3/envs/pypeit/lib/python3.11/site-packages/pypeit/scripts/scriptbase.py", line 113, in entry_point
    cls.main(cls.parse_args())
  File "/opt/anaconda3/envs/pypeit/lib/python3.11/site-packages/pypeit/scripts/sensfunc.py", line 185, in main
    sensobj.run()
  File "/opt/anaconda3/envs/pypeit/lib/python3.11/site-packages/pypeit/sensfunc.py", line 453, in run
    self.flux_std()
  File "/opt/anaconda3/envs/pypeit/lib/python3.11/site-packages/pypeit/sensfunc.py", line 467, in flux_std
    self.sobjs_std.apply_flux_calib(self.par_fluxcalib, self.spectrograph, self)
  File "/opt/anaconda3/envs/pypeit/lib/python3.11/site-packages/pypeit/specobjs.py", line 576, in apply_flux_calib
    if par['extinct_correct'] is None else par['extinct_correct']
       ~~~^^^^^^^^^^^^^^^^^^^
TypeError: 'NoneType' object is not subscriptable

Still meet the error.

**Comment by debora-pe** [comment]
Hi @Codeprey. Thank you for testing this. I found a quick fix to this bug. If you can, please take a look at PR #1797  and test it with your data by using the `fix_parfluxcalib` branch. Thanks!

**Comment by Codeprey** [comment]
Thanks Debora, I cannot directly install the fix_parfluxcalib branch since the pip install insists that the branch has no version number. But I have switched the sensfunc.py with your new script. This problem is now gone!

However, I meet another problem in the telluric absorption correction. I have tested on the old TelFit_Lick_3100_11100_R10000.fits (5.8GB) grid file [used on v1.6.0, which is quite nice], and also downloaded the new grid file TelFit_Palomar_9000_25000_R10000.fits (5.9GB) [don't exactly know if it should be used, but I guess it should be great for P200 DBSP]. 

In both cases, I encounter this bug:
Input:
pypeit_tellfit new_red.fits -t p200_dbsp_red.tell -g TelFit_Palomar_9000_25000_R10000.fits [TelFit_Lick_3100_11100_R10000.fits]

Output:
[INFO]    :: Loading the reduction file
[INFO]    :: Telluric input file loaded successfully.
[INFO]    :: Writing the telluric fitting parameters to telluric.par
[WARNING] :: Selected configuration file already exists and will be overwritten!
[INFO]    :: Telluric-corrected spectrum will be saved to: new_red_tellcorr.fits.
[INFO]    :: Best-fit telluric model will be saved to: new_red_tellmodel.fits.
Traceback (most recent call last):
  File "/opt/anaconda3/envs/pypeit/bin/pypeit_tellfit", line 8, in <module>
    sys.exit(TellFit.entry_point())
             ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda3/envs/pypeit/lib/python3.11/site-packages/pypeit/scripts/scriptbase.py", line 113, in entry_point
    cls.main(cls.parse_args())
  File "/opt/anaconda3/envs/pypeit/lib/python3.11/site-packages/pypeit/scripts/tellfit.py", line 201, in main
    TelPoly = telluric.poly_telluric(args.spec1dfile, par['telluric']['telgridfile'],
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda3/envs/pypeit/lib/python3.11/site-packages/pypeit/core/telluric.py", line 1864, in poly_telluric
    TelObj = Telluric(wave, flux, ivar, mask_tot, telgridfile, obj_params, init_poly_model,
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda3/envs/pypeit/lib/python3.11/site-packages/pypeit/core/telluric.py", line 2420, in __init__
    self.tell_dict = read_telluric_pca(self.telgrid, wave_min=self.wave_in_arr[wv_gpm].min(),
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda3/envs/pypeit/lib/python3.11/site-packages/pypeit/core/telluric.py", line 178, in read_telluric_pca
    ncomp = hdul[0].header['NCOMP']
            ~~~~~~~~~~~~~~^^^^^^^^^
  File "/opt/anaconda3/envs/pypeit/lib/python3.11/site-packages/astropy/io/fits/header.py", line 170, in __getitem__
    card = self._cards[self._cardindex(key)]
                       ^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda3/envs/pypeit/lib/python3.11/site-packages/astropy/io/fits/header.py", line 1726, in _cardindex
    raise KeyError(f"Keyword {keyword!r} not found.")
KeyError: "Keyword 'NCOMP' not found."


When a default grid TellPCA_3000_26000_R10000.fits (6.3MB) is used, there is no such a bug. However, the correction is not as good as expected, which means the much larger grid file(s) should be used.

But how can I avoid this bug? I am really confused. According to the document, the 'NCOMP' parameter is the 'Number of PCA components'. I really appreciate it if you can give me some suggestions, which may also be useful for other P200 spectrum reducers!

Best,
Peter

**Comment by debora-pe** [comment]
@Codeprey if you don't mind, I'll keep this issue focused on the original problem that you encounter. I will close it as soon as the PR is cleared.

I am not very familiar with how the telluric grid files are generated or used. I would suggest asking on the [PypeIt Users Slack](https://pypeit-users.slack.com/). You will find somebody that can help you.

**Comment by debora-pe** [comment]
@Codeprey have you tried setting the parameter `teltype=grid` in you .tell configuration file (the default is pca)? See https://pypeit.readthedocs.io/en/latest/pypeit_par.html#telluricpar-keywords

**Comment by debora-pe** [comment]
This was resolved in PR #1797 . Closing this issue.
